sapi/fpm/fpm/fpm_unix.c

```diff
@@ -392,6 +402,105 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
 }
 /* }}} */
 
+#if HAVE_FPM_CPUAFFINITY
+struct fpm_cpuaffinity_conf {
+#if defined(HAVE_SCHED_SETAFFINITY) || defined(HAVE_CPUSET_SETAFFINITY)
+	cpu_set_t cset;
+#endif
+	long min;
+	long max;
+};
+
+static long fpm_cpumax(void)
+{
+	static long cpuid = LONG_MIN;
+	if (cpuid == LONG_MIN) {
+		cpu_set_t cset;
+#if defined(HAVE_SCHED_SETAFFINITY)
+		if (sched_getaffinity(0, sizeof(cset), &cset) == 0) {
+#elif defined(HAVE_CPUSET_SETAFFINITY)
+		if (cpuset_getaffinity(CPU_LEVEL_WHICH, CPU_WHICH_PID, -1, sizeof(cset), &cset) == 0) {
+#endif
+			cpuid = CPU_COUNT(&cset);
+		} else {
+			cpuid = -1;
+		}
+	}
+
+	return cpuid;
+}
+
+static void fpm_cpuaffinity_init(struct fpm_cpuaffinity_conf *c)
+{
+#if defined(HAVE_FPM_CPUAFFINITY)
+	CPU_ZERO(&c->cset);
+#endif
+}
+
+static void fpm_cpuaffinity_add(struct fpm_cpuaffinity_conf *c)
+{
+#if defined(HAVE_FPM_CPUAFFINITY)
+	int i;
+
+	for (i = c->min; i <= c->max; i ++) {
+		if (!CPU_ISSET(i, &c->cset)) {
+			CPU_SET(i, &c->cset);
+		}
+	}
+#endif
+}
+
+static int fpm_cpuaffinity_set(struct fpm_cpuaffinity_conf *c)
+{
+#if defined(HAVE_SCHED_SETAFFINITY)
+	return sched_setaffinity(0, sizeof(c->cset), &c->cset);
+#elif defined(HAVE_CPUSET_SETAFFINITY)
+	return cpuset_setaffinity(CPU_LEVEL_WHICH, CPU_WHICH_PID, -1, sizeof(c->cset), &c->cset);
+#endif
+}
+
+static int fpm_setcpuaffinity(char *cpu_list)
+{
+	char *token, *buf;
+	struct fpm_cpuaffinity_conf fconf;
+	int r, cpumax;
+
+	r = -1;
+	cpumax = fpm_cpumax();
+
+	fpm_cpuaffinity_init(&fconf);
+	token = php_strtok_r(cpu_list, ",", &buf);
+
+	while (token) {
+		char *cpu_listsep;
+
+		fconf.min = strtol(token, &cpu_listsep, 0);
+		if (errno || fconf.min < 0 || fconf.min > cpumax) {
+			return -1;
+		}
+		fconf.max = fconf.min;
+		if (*cpu_listsep == '-') {
+			if (strlen(cpu_listsep) > 1) {
+				char *err;
+				fconf.max = strtol(cpu_listsep + 1, &err, 0);
+				if (errno || *err != '\0' || fconf.max < fconf.min || fconf.max > cpumax) {
+					return -1;
+				}
```

Comment:
This is currently not covered by any test in the same way as the list - there should be some tests that would run if more cpus available (doesn't matter that they might not run in the pipeline - they will run on my computer at least)